### DISCUSSION TITLE
general question about relocatable buildroot

### DISCUSSION THREAD

**Discussion by pellcorp** [xblax/flashforge_ad5m_klipper_mod#209] [General]
Loving this mod, although I am only using it for parts rather than for its original intention.   

I did consider looking at entware or even building my own packages directly, but buildroot is awesome for making it far easier to create a complete collection of packages and dependent libraries, the only thing its missing to make it truly awesome for my use case is for build rootfs.tar to be useable on an existing rootfs with different libc.

I really just want to use this project something like entware does (packages going to /opt), but I found buildroot so much simpler to understand and tweak, and also I forked ballasway version of this project, and he has already done all the heavy lifting to get it to work with MIPS, so I wanted to take advantage of that as well.

I am using patchelf to change the interpreter and the rpath of runnable binaries and python so files.   So far its early days and I a imagine I will run into issues.

But I am wondering is there a better way to be able to use the output of buildroot to supplement rather than replace an existing rootfs?

I also discovered that buildroot patchelf does nasty things to MIPS libaries (well only set-rpath does), but that using Ubuntu 22.04 or debian bookworm patchelf works fine.

**Comment by xblax** [comment]
Well the main idea behind this mod's approach is to use the buildroot system in a chroot-environment. In this scenario the libc does not really matter, because the chroot environment brings it's own libc and mainly the kernel headers are relevant.

Other approaches get complicated quickly. In general you can target an old libc version for the build to be compatible with newer versions, so you would probably compile against the oldest version that the applications still support. 

If you don't want to use a chroot but set the rpath of all binaries instead, can't you also have your own libc then? I've never tried this, but it's an interesting idea.

**Reply by pellcorp** [reply]
Ya I just deploy the complete rootfs.tar to /opt, including the new libc built by buildroot 

So far preliminary tests look good but I am still very cautious about long term success 😀

So because I am basing my stuff on what ballaswag is already doing the kernel headers are the compatible version already.

**Reply by pellcorp** [reply]
If any of the binaries rely on --prefix in addition to standard lib dir locations (which I think I can solve) I may run into issues, so for instance nginx has some hard coded stuff in the make file referencing /var/tmp and of course a default config file being in /etc.  but if it does not get any more tricky than that things might just work 😀

I mean in this case I could just fork buildroot to tweak the .mk files in that case to point at /opt ... Hopefully I don't have to though
